### PR TITLE
Java: Add `FCALL` command.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
https://redis.io/docs/latest/commands/fcall/

**Some details**
common interface functions
```js
Object fcall(String function, String[] keys, String[] arguments);
```
standalone interface functions
```js
Object fcall(String function);
```
cluster interface functions
```js
Object fcall(String function);
Object fcall(String function, Route route);
Object fcall(String function, String[] arguments);
Object fcall(String function, String[] arguments, Route route);
```

**This results in**

standalone client functions
```js
Object fcall(String function);
Object fcall(String function, String[] keys, String[] arguments);
```
cluster client functions
```js
Object fcall(String function); // routed to a random node
Object fcall(String function, Route route); // routed according to the given route
Object fcall(String function, String[] arguments); // routed to a random node
Object fcall(String function, String[] arguments, Route route); // routed according to the given route
Object fcall(String function, String[] keys, String[] arguments); // routed according to the given keys
```

transaction functions
```js
Object fcall(String function, String[] keys, String[] arguments);
Object fcall(String function, String[] arguments);
```

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 1]
```diff
@@ -84,4 +84,41 @@ public interface ScriptingAndFunctionsClusterCommands {
...
+     *
+     * @since Redis 7.0 and above.
+     * @see <a href="https://redis.io/docs/latest/commands/fcall/">redis.io</a> for details.
+     * @param function The function name.
+     * @param arguments An <code>array</code> of <code>function</code> arguments.
+     * @return A value depends on the function that was executed.
+     * @example
+     *     <pre>{@code
+     * String[] args = new String[] { "Answer", "to", "the", "Ultimate", "Question", "of", "Life,", "the", "Universe,", "and", "Everything"};
+     * Object response = client.fcall("Deep_Thought", args).get();
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:100]: 🤔

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     *       <li>all <code>keys</code> must map to the same hash slot.
+     *       <li>and no <code>keys</code> are given, command will be routed to a random node.
+     *     </ul>
+     *
+     * @since Redis 7.0 and above.
+     * @see <a href="https://redis.io/docs/latest/commands/fcall/">redis.io</a> for details.
+     * @param function The function name.
+     * @param keys An <code>array</code> of key names which <code>function</code> will work with.
+     * @param arguments An <code>array</code> of <code>function</code> arguments.
+     * @return A value depends on the function that was executed.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java:29]: What do you think of this? 
```suggestion
     * @return The invoked function's return value.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+public interface ScriptingAndFunctionsBaseCommands {
+
+    /**
+     * Invokes a previously loaded function.
+     *
+     * @apiNote When in cluster mode
+     *     <ul>
+     *       <li>all <code>keys</code> must map to the same hash slot.
+     *       <li>and no <code>keys</code> are given, command will be routed to a random node.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java:21]:
```suggestion
     *       <li>if no <code>keys</code> are given, command will be routed to a random node.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+     *     </ul>
+     *
+     * @since Redis 7.0 and above.
+     * @see <a href="https://redis.io/docs/latest/commands/fcall/">redis.io</a> for details.
+     * @param function The function name.
+     * @param keys An <code>array</code> of key names which <code>function</code> will work with.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java:27]:
```suggestion
     * @param keys An <code>array</code> of key arguments accessed by the function.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+     * @param arguments An <code>array</code> of <code>function</code> arguments.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java:28]:
```suggestion
     * @param arguments An <code>array</code> of non-key input arguments accessed by the function.
```

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 6]
```diff
@@ -524,11 +524,15 @@ private static Object[] scriptingAndFunctionsCommands(BaseTransaction<?> transac
                 "#!lua name=mylib1T \n"
                         + " redis.register_function('myfunc1T', function(keys, args) return args[1] end)";
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:525]: any chance you could update the function to access and use a key?  that way we would know that they're being passed successfully.

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 7]
```diff
@@ -689,7 +690,13 @@ public static Stream<Arguments> callCrossSlotCommandsWhichShouldFail() {
                 Arguments.of(
                         "bzpopmin", "5.0.0", clusterClient.bzpopmin(new String[] {"abc", "zxy", "lkn"}, .1)),
                 Arguments.of(
-                        "bzmpop", "7.0.0", clusterClient.bzmpop(new String[] {"abc", "zxy", "lkn"}, MAX, .1)));
+                        "bzmpop",
+                        "7.0.0",
+                        clusterClient.bzmpop(new String[] {"abc", "zxy", "lkn"}, MAX, .1),
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:695]:
```suggestion
                        clusterClient.bzmpop(new String[] {"abc", "zxy", "lkn"}, MAX, .1)),
```

File: java/integTest/src/test/java/glide/cluster/CommandTests.java [hunk 8]
```diff
@@ -784,16 +798,49 @@ public void functionLoad(boolean withRoute) {
...
+                        + newFuncName
                         + "', function(keys, args) return #args end)";
         promise2 =
                 withRoute
                         ? clusterClient.functionLoadReplace(newCode, route)
                         : clusterClient.functionLoadReplace(newCode);
         assertEquals(libName, promise2.get());
-        // TODO test with FCALL
+
+        Thread.sleep(1000);
```
Comment (acarbonetto) [inline java/integTest/src/test/java/glide/cluster/CommandTests.java:813]: why?

File: java/integTest/src/test/java/glide/cluster/CommandTests.java (hunk 7 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:695]: Oh, how it worked?

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:525]: Need to add a separate test - common transaction test considered to be keyless.
Will do

File: java/integTest/src/test/java/glide/cluster/CommandTests.java (hunk 8 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/cluster/CommandTests.java:813]: GC fail

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @see <a href="https://redis.io/docs/latest/commands/?group=scripting">Scripting and Function
+ *     Commands</a>
+ */
+public interface ScriptingAndFunctionsBaseCommands {
+
+    /**
+     * Invokes a previously loaded function.
+     *
+     * @apiNote When in cluster mode
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsBaseCommands.java:18]: add `Object fcall(String)` signature

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 10]
```diff
@@ -84,4 +84,74 @@ public interface ScriptingAndFunctionsClusterCommands {
      * }</pre>
      */
     CompletableFuture<String> functionLoadReplace(String libraryCode, Route route);
+
+    /**
+     * Invokes a previously loaded function.<br>
```
Comment (yipin-chen) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:89]: extra <br>?

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:89]: No, it is intentionally here to put phrase about routing on the next line

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java]: Added in fcc4b7a8

**Review by GumpacG — APPROVED** [review]

**Pull request merged**